### PR TITLE
Add proposal for lookups

### PR DESCRIPTION
**Pull request by iwootten**
### What is the context of this PR?

This proposes how lookups might be represented in a survey schema. 

### How to review 

- See if you think the proposal makes sense.
- Is there anything I've missed?

### CODE REVIEW

**Comment by iwootten** [comment]
I had pondered whether including a limit for the number of results to display might be appropriate?

**Comment by bitdivision** [comment]
I think when this was initially discussed, we talked about having the ability to avoid using a lookup api for simple lookups, i.e. where there are not many options. Obviously there are some disadvantages to this, so I assume this was ruled out?

**Comment by iwootten** [comment]
@bitdivision - don't remember that conversation, but I guess that's still an option if we set a flag to indicate the query_path should be used against the runner host itself?

**Comment by ajmaddaford** [comment]
@bitdivision I think the intent of the lookup has changed as it's gone through user testing. The current design/prototype is for the suggestions from the lookup to behave more like a search than an autocomplete. If we weren't using the api we'd end up having to have two implementations of that, one client side and another in the api.

File: doc/architecture/decisions/0008-lookups.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            id: 'religion-answer',
+            mandatory: false,
+            label: 'Religion',
+            type: 'Lookup',
+            query_path: 'path/to/query'
+          }
+        ]
+    }
+
+The answer behaves as a Textfield when JavaScript is disabled and specifies a query_path for the frontend component to retrieve suggestions from when enabled.
```
Comment (ajmaddaford) [inline doc/architecture/decisions/0008-lookups.md]: I think it behaves as a Textfield regardless of Javascript. The Javascript will add the suggestions list below the textfield.

image

**Comment by pricem14pc** [comment]
> I think when this was initially discussed, we talked about having the ability to avoid using a lookup api for simple lookups, i.e. where there are not many options. Obviously there are some disadvantages to this, so I assume this was ruled out?

Presumably all the look ups need to behave the same regardless of the number of entries in the datasets? Unless a simple lookup could replicate the behaviour of an elasticsearch api (e.g. partial tokenized matches etc) then it might impact the user experience.

**Review by ajmaddaford — APPROVED** [review]

**Review by bitdivision — APPROVED** [review]

**Review by MebinAbraham — CHANGES_REQUESTED** [review]
>I had pondered whether including a limit for the number of results to display might be appropriate?

Do we not want to do add an option for this, instead of leaving it to the front end to determine a hard limit on all look ups?.

Is the only reason we may not want it to be configurable in the schema is that it's researched and the current hard limit just works and wouldn't need changing?

@iwootten @ajmaddaford @bitdivision

File: doc/architecture/decisions/0008-lookups.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            id: 'religion-answer',
+            mandatory: false,
+            label: 'Religion',
+            type: 'Lookup',
+            query_path: 'path/to/query'
+          }
+        ]
+    }
+
+The answer behaves as a Textfield and specifies a query_path for the frontend component to retrieve suggestions from when JavaScript is enabled.
```
Comment (MebinAbraham) [inline doc/architecture/decisions/0008-lookups.md]: 🤷‍♂ Personally prefer if variables in sentences are highlighted in some way.
```suggestion
The answer behaves as a Textfield and specifies a `query_path` for the frontend component to retrieve suggestions from when JavaScript is enabled.
```

File: doc/architecture/decisions/0008-lookups.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+The api to query is a service that takes an input string and returns a single flat array of suggestion strings. 
+
+Initially there will be no authentication.
+
+One service is expected to host all single tier lookups.
+
+The hostname of the service is specified as an environment variable.
+
+The hostname would be combined with the query_path attribute at runtime to determine the full api url.
```
Comment (MebinAbraham) [inline doc/architecture/decisions/0008-lookups.md]:
```suggestion
The hostname would be combined with the `query_path` attribute at runtime to determine the full api url.
```

**Comment by MebinAbraham** [comment]
Curious, does `query_path` need to live in the schema?. Have we considered a `lookup_type` option instead?. Means it would only need to live in runner and validator wouldn't need knowledge of the path itself and that's just another place to update when/if the path changes.

**Comment by bitdivision** [comment]
@MebinAbraham Regarding the number of items shown, we discussed this briefly yesterday and I think the conclusion was that there should be one number which makes sense for all devices and that number can be chosen by the product owner. There shouldn't ever be a need to modify it on a per schema basis.

**Review by MebinAbraham — APPROVED** [review]

**Review by bitdivision — APPROVED** [review]

**Pull request merged**